docs: Update migration guide (#224)
kyma-project/modulectl: docs/contributor/migration-guide.md

repository: https://github.com/kyma-project/<module-manager-name>.git
version: 1.34.0
manifest: https://github.com/kyma-project/<module-manager>/releases/download/1.34.0/<module-manager-name>.yaml
defaultCR: https://github.com/kyma-project/<module-manager>/releases/download/1.34.0/<module-name-default-cr>.yaml
security: sec-scanners-config.yaml
manager:
   name: <module-manager-name>
   namespace: kyma-system
   group: apps
   version: v1
   kind: Deployment
associatedResources:
   - group: operator.kyma-project.io
     kind: <ModuleName>
     version: v1alpha1
   - group: operator.kyma-project.io
     kind: LogParser
     version: v1alpha1
   - group: operator.kyma-project.io
     kind: LogPipeline
     version: v1alpha1
   - group: operator.kyma-project.io
     kind: MetricPipeline
     version: v1alpha1
   - group: operator.kyma-project.io
     kind: TracePipeline
     version: v1alpha1
documentation: https://help.sap.com/docs/btp/sap-business-technology-platform/<kyma-module-name>
icons:
   - name: module-icon
     link: https://raw.githubusercontent.com/kyma-project/kyma/refs/heads/main/docs/assets/logo_icon.svg
```

## 3. ModuleTemplate Differences 

> The following table compares the key fields in the `ModuleTemplate` YAML generated by Kyma CLI vs. modulectl.

Reviewer: ```suggestion
### 3.1 Field Mapping Differences

The following table compares the key fields in the `ModuleTemplate` YAML generated by Kyma CLI vs. modulectl.
```

| **spec.manager.group**                                              | *n/a*                                                | **New**                                                     |
| **spec.manager.version**                                            | *n/a*                                                | **New**                                                     |
| **spec.data**                                                       | Present                                              | Unchanged                                                   |
| **spec.info**                                                       | *n/a*                                                | **New**                                                     |
| **spec.descriptor.component**                                       | Present                                              | Unchanged                                                   |
| **spec.resources**                                                  | *n/a*                                                | **New**                                                     |

> See the following examples of `ModuleTemplate` YAML files generated for TemplateOperator by Kyma CLI vs. modulectl.

Reviewer: ```suggestion
### 3.2 Examples

See the following examples of `ModuleTemplate` YAML files generated for TemplateOperator by Kyma CLI vs. modulectl.
```

`modulectl` is the successor of the module developer-facing capabilities of Kyma CLI.
It is already tailored for the updated ModuleTemplate metadata as discussed in [ADR: Iteratively moving forward with module requirements and aligning responsibilities](https://github.com/kyma-project/lifecycle-manager/issues/1681).

## 1. Tooling & Workflow Changes

This section focuses on the `modulectl` CLI itself and the related submission, deployment, and migration workflows.

### 1.1 Use `modulectl`
Modulectl is available for download from the [GitHub Releases](https://github.com/kyma-project/modulectl/releases).
For an overview of the supported commands and flags, use `modulectl -h` or `modulectl <command> -h` to show the definitions.

```bash
modulectl -h                # general help
modulectl create -h         # help for 'create'
modulectl scaffold -h       # help for 'scaffold'
```

### 1.2 Command & Flag Differences

#### 1.2.1 Command Mapping

| Operation                | Kyma CLI                       | `modulectl`                         |
|--------------------------|--------------------------------|-------------------------------------|

> | Scaffold module template | N/A                            | `modulectl scaffold ...`            |

Reviewer 2: I think `scaffold` was also supported in CLI

| ----------------------------- | ------------------------------------------------- | -------------------------------------------------------------- |
| Config-file flag              | `--module-config-file`                            | `--config-file`, shortcut `-c` (all commands)                  |
| Archive overwrite             | `--module-archive-version-overwrite`              | **Removed** (module versions are immutable)                    |
| ModuleTemplate naming pattern | `.metadata.name = <module-name>-<channel>`        | `.metadata.name = <module-name>-<version>`                     |
| Version label                 | `operator.kyma-project.io/module-version` label   | Populated in `.spec.version`                                   |
| Documentation annotation      | `operator.kyma-project.io/doc-url` annotation     | Defined in `.spec.info.documentation` via `documentation:` key |
| Manifest & DefaultCR source   | Local file references only                        | Fetched directly from GitHub release URLs                      |

> | Release channel in config     | Required `.channel` field in `module-config.yaml` | **Removed**; channel mapping managed by ModuleReleaseMeta CR   |

Reviewer 2: I think I would keep this secton about flags only, and move these differences in the module-config and the generated ModuleTemplate to `2.2` and the new section proposed below.

3. **Update Channel Mapping**: Edit `/modules/<module>/module-releases.yaml` to map channels to versions.
4. **Trigger Pipeline**: On PR merge, the submission pipeline:

   * Validates schema, FQDN, version uniqueness
   * Builds via `modulectl`, pushes OCI image
   * Generates ModuleTemplate and ModuleReleaseMeta in `/kyma/kyma-modules`
5. **ArgoCD Sync**: ArgoCD deploys ModuleTemplates and ModuleReleaseMeta to KCP landscapes.
6. **Cleanup**: Remove obsolete channel-based configs when stable.

#### Migration Period & Coexistence

* Both channel-based (old) and version-based (new) metadata are supported during the transition.
* KLM reads version-based templates first; falls back to channel-based if missing.
* After full migration, legacy `<module>-<channel>` templates are rejected.

---
## 2. Module Configuration & Metadata Changes

This section covers the structure and content of `module-config.yaml` and `module-releases.yaml` files under the new version-based layout.

### 2.1 Directory Structure

* **Old Layout**: `/modules/<module-name>/<channel>/module-config.yaml`
* **New Layout**: `/modules/<module-name>/<version>/module-config.yaml`

> ### 2.2 `module-config.yaml` Schema Comparison Example

Reviewer 2: Also I think we should add another section about the generated ModuleTemplate and also list the changes there.

- list the newly introduced fields
- changed/moved fields, e.g., docs-url label now spec.documentation, version label now spec.version, name now `<module>-<version>` instead of `<module>-<channel>`, ...

security: sec-scanners-config.yaml
manager:
   name: <module-manager-name>
   namespace: kyma-system
   group: apps
   version: v1
   kind: Deployment
associatedResources:
   - group: operator.kyma-project.io
     kind: <ModuleName>
     version: v1alpha1
   - group: operator.kyma-project.io
     kind: LogParser
     version: v1alpha1
   - group: operator.kyma-project.io
     kind: LogPipeline
     version: v1alpha1
   - group: operator.kyma-project.io
     kind: MetricPipeline
     version: v1alpha1
   - group: operator.kyma-project.io
     kind: TracePipeline
     version: v1alpha1
documentation: https://help.sap.com/docs/btp/sap-business-technology-platform/<kyma-module-name>
icons:
   - name: module-icon
      # TODO: provide <module-name> icon
     link: https://raw.githubusercontent.com/kyma-project/kyma/refs/heads/main/docs/assets/logo_icon.svg
```

### 2.3 Channel Mapping with ModuleReleaseMeta Example

```yaml
# modules/<module-name>/module-releases.yaml
channels:
  - channel: regular
    version: 1.34.0
  - channel: fast
    version: 1.34.0
  - channel: experimental
    version: 1.34.0-experimental
  - channel: dev
    version: 1.35.0-rc1
```

On merge, the pipeline:

> * Validates no downgrades and version existence

Reviewer 2: I think we could trim this section entirely. Good to mention that the channel mapping is handled in another file now. But since modulectl is not involved there no need to mention this here. This is part of the https://github.com/kyma-project/lifecycle-manager/pull/2452

* Generates ModuleReleaseMeta CRs per landscape
* Updates landscape-specific kustomizations to reference only active versions

### 2.4 Metadata Deprecations & New Practices

| Deprecated Feature                | Replacement / New Location                                                    |
| --------------------------------- | ----------------------------------------------------------------------------- |
| `.channel` field in module config | Moved to ModuleReleaseMeta CR (`module-releases.yaml`)                        |

> | `mandatory` on ModuleTemplate     | Set `mandatory: true` in module config; reconciler picks highest version      |

Reviewer 3: Reconciler picks highest version? Is this meaning lifecycle-manager? Doesn't it try to always install the highest version?

Reviewer 2: Also not sure if this is needed in this doc.

`modulectl` is the successor of the module developer-facing capabilities of Kyma CLI.
It is already tailored for the updated ModuleTemplate metadata as discussed in [ADR: Iteratively moving forward with module requirements and aligning responsibilities](https://github.com/kyma-project/lifecycle-manager/issues/1681).

## 1. Tooling & Workflow Changes

This section focuses on the `modulectl` CLI itself and the related submission, deployment, and migration workflows.

### 1.1 Use `modulectl`

> It is available for download from the [GitHub Releases](https://github.com/kyma-project/modulectl/releases).

Reviewer: ```suggestion
Modulectl is available for download from the [GitHub Releases](https://github.com/kyma-project/modulectl/releases).
```

For an overview of the supported commands and flags, use `modulectl -h` or `modulectl <command> -h` to show the definitions.

```bash
modulectl -h                # general help
modulectl create -h         # help for 'create'
modulectl scaffold -h       # help for 'scaffold'
modulectl list -h           # help for 'list'
```

### 1.2 Command & Flag Differences

#### 1.2.1 Command Mapping

| Operation                  | Kyma CLI                 | `modulectl`                              |
| -------------------------- | ------------------------ |------------------------------------------|
| Scaffold module template   | `kyma module init ...`   | `modulectl scaffold ...`                 |
| Create ModuleTemplate CR   | `kyma module create ...` | `modulectl create -c <config-file>`      |

> | List templates or versions | `kyma module list ...`   | \`modulectl list \[templates releases]\` |

Reviewer: Why is the `modulectl list` command formatted differently? Do you mean that you can list either templates or releases?

```suggestion
| List templates or versions | `kyma module list ...`   | `modulectl list templates`<br>`modulectl list releases` |
```

Author: That's Right , need to be deleted as I double checked once again , there's no listing command on `modulectl`

| Config-file flag              | `--module-config-file`                            | `--config-file`, shortcut `-c` (all commands)                  |
| Archive overwrite             | `--module-archive-version-overwrite`              | **Removed** (module versions are immutable)                    |
| ModuleTemplate naming pattern | `.metadata.name = <module-name>-<channel>`        | `.metadata.name = <module-name>-<version>`                     |
| Version label                 | `operator.kyma-project.io/module-version` label   | Populated in `.spec.version`                                   |
| Documentation annotation      | `operator.kyma-project.io/doc-url` annotation     | Defined in `.spec.info.documentation` via `documentation:` key |
| Manifest & DefaultCR source   | Local file references only                        | Fetched directly from GitHub release URLs                      |
| Release channel in config     | Required `.channel` field in `module-config.yaml` | **Removed**; channel mapping managed by ModuleReleaseMeta CR   |

### 1.3 Submission Process & Migration Period

> #### Submission Process

Reviewer: ```suggestion
#### Module Submission Process
```
Do I understand that this is about how new module versions are submitted?

1. **Publish Artifacts**: Release module version assets (manifest, defaultCR) on GitHub.
2. **Submit Version Config**: Add/update `/modules/<module>/<version>/module-config.yaml`.
3. **Update Channel Mapping**: Edit `/modules/<module>/module-releases.yaml` to map channels to versions.
4. **Trigger Pipeline**: On PR merge, the submission pipeline:

   * Validates schema, FQDN, version uniqueness
   * Builds via `modulectl`, pushes OCI image
   * Generates ModuleTemplate and ModuleReleaseMeta in `/kyma/kyma-modules`

> 5. **ArgoCD Sync**: Deploys ModuleTemplates and ModuleReleaseMeta to KCP landscapes.

Reviewer: Who is responsible for that? Should we write "Deploy ModuleTemplates..." as this is done by the module team or is it also done on the PR merge automatically?

Author: After the module is promoted , ArgoCD picks changes made by submission pipeline and deploys the ModuleTemplates and ModuleReleaseMetas relevant for the respective landscape.

Reviewer: ```suggestion
5. **ArgoCD Sync**: ArgoCD deploys ModuleTemplates and ModuleReleaseMeta to KCP landscapes.
```

Reviewer: Maybe we can add the subject explicitly then?

6. **Cleanup**: Remove obsolete channel-based configs when stable.

#### Migration Period & Coexistence

> * Both channel-based (old) and version-based (new) metadata are supported during the transition.

Reviewer: Do we already how long the transition phase is?

Reviewer: Should we mention the timeline in the guide?

    version: 1.34.0-experimental
  - channel: dev
    version: 1.35.0-rc1
```

On merge, the pipeline:

* Validates no downgrades and version existence
* Generates ModuleReleaseMeta CRs per landscape
* Updates landscape-specific kustomizations to reference only active versions

### 2.4 Metadata Deprecations & New Practices

| Deprecated Feature                | Replacement / New Location                                                    |
| --------------------------------- | ----------------------------------------------------------------------------- |
| `.channel` field in module config | Moved to ModuleReleaseMeta CR (`module-releases.yaml`)                        |
| `mandatory` on ModuleTemplate     | Set `mandatory: true` in module config; reconciler picks highest version      |
| Beta/Internal flags on templates  | Configured in ModuleReleaseMeta via `.spec.info.beta` / `.spec.info.internal` |

---

## Additional Resources

- [`modulectl` GitHub Repository](https://github.com/kyma-project/modulectl)
- [ADR: Iteratively moving forward with module requirements and aligning responsibilities](https://github.com/kyma-project/lifecycle-manager/issues/1681)

> - [ModuleTemplate CR Reference](https://github.com/kyma-project/lifecycle-manager/blob/main/docs/contributor/resources/03-moduletemplate.md)
> - [ModuleReleaseMeta CR Reference](https://github.com/kyma-project/lifecycle-manager/blob/main/docs/contributor/resources/05-modulereleasemeta.md)

Reviewer: ```suggestion
- [ModuleTemplate Custom Resource](https://github.com/kyma-project/lifecycle-manager/blob/main/docs/contributor/resources/03-moduletemplate.md)
- [ModuleReleaseMeta Custom Resource](https://github.com/kyma-project/lifecycle-manager/blob/main/docs/contributor/resources/05-modulereleasemeta.md)
```

`modulectl` is the successor of the module developer-facing capabilities of Kyma CLI.
It is already tailored for the updated ModuleTemplate metadata as discussed in [ADR: Iteratively moving forward with module requirements and aligning responsibilities](https://github.com/kyma-project/lifecycle-manager/issues/1681).

> ## 1. Tooling & Workflow Changes

Reviewer: Is this section still about the workflow?

Reviewer: Ok.

> This section focuses on the `modulectl` CLI itself and the related submission, deployment, and migration workflows.

Reviewer: ```suggestion
This section focuses on the modulectl CLI itself and the related submission, deployment, and migration workflows.
```
Mo, are the submission, deployment, and migration workflows still described here?

| `moduleRepoTag`                                | `moduleRepoTag`                  | Indicator for the pipeline to checkout the provided tag (default tag to checkout is the `version`)                                                                          |
| `beta`                                         | **Removed**                      | Marks the module as beta. Moved to [ModuleReleaseMeta](https://github.com/kyma-project/lifecycle-manager/blob/main/docs/contributor/resources/05-modulereleasemeta.md)                                                                                                                                |
| `internal`                                     | **Removed**                      | Marks the module as internal. Moved to [ModuleReleaseMeta](https://github.com/kyma-project/lifecycle-manager/blob/main/docs/contributor/resources/05-modulereleasemeta.md)                                                                                                                    |

## 2.2 Example:

> ### 2.2.1 Module Config using Kyma CLI

Reviewer: Bold should also be enough for listing the examples; it does not seem to require a new subsection.

| **spec.manager.group**                                              | *n/a*                                                | **New**                                                     |
| **spec.manager.version**                                            | *n/a*                                                | **New**                                                     |
| **spec.data**                                                       | Present                                              | Unchanged                                                   |
| **spec.info**                                                       | *n/a*                                                | **New**                                                     |
| **spec.descriptor.component**                                       | Present                                              | Unchanged                                                   |
| **spec.resources**                                                  | *n/a*                                                | **New**                                                     |

> The following shows the `ModuleTemplate` YAML generated for TemplateOperator by Kyma CLI vs. `modulectl`.

Reviewer: ```suggestion
See the following examples of `ModuleTemplate` YAML files generated for TemplateOperator by Kyma CLI vs. modulectl.
```

## 3.1 Kyma CLI–Generated ModuleTemplate (channel-based)

```yaml
apiVersion: operator.kyma-project.io/v1beta2
kind: ModuleTemplate
metadata:
  name: template-operator-regular
  namespace: kcp-system
  labels:
    "operator.kyma-project.io/module-name": "template-operator"
  annotations:
    "operator.kyma-project.io/module-version": "0.0.1"
spec:
  mandatory: false
  channel: regular
  data:
    apiVersion: operator.kyma-project.io/v1alpha1
    kind: Sample
    metadata:
      name: sample-yaml
    spec:
      resourceFilePath: "./module-data/yaml"
  descriptor:
    component:
      # ... component descriptor
    meta:
      schemaVersion: v2
```

> ## 3.2 `modulectl`–Generated ModuleTemplate (version-based)

Reviewer: ```suggestion
**modulectl–generated ModuleTemplate (version-based)**
```

  name: template-operator-0.0.1
  namespace: kcp-system
  resourceVersion: "4369227537"
  uid: b00c8fcc-aaa2-43ce-bdab-ba82365e11b7
spec:
  associatedResources:
  - group: operator.kyma-project.io
    kind: Managed
    version: v1alpha1
  data:
    apiVersion: operator.kyma-project.io/v1alpha1
    kind: Sample
    metadata:
      name: sample-yaml
    spec:
      resourceFilePath: ./module-data/yaml
  descriptor:
    component:
      # ... component descriptor
    meta:
      schemaVersion: v2
  info:
    documentation: https://github.com/kyma-project/template-operator/blob/main/README.md
    icons:
    - link: https://raw.githubusercontent.com/kyma-project/kyma/refs/heads/main/docs/assets/logo_icon.svg
      name: module-icon
    repository: https://github.com/kyma-project/template-operator
  manager:
    group: apps
    kind: Deployment
    name: template-operator-controller-manager
    namespace: template-operator-system
    version: v1
  mandatory: false
  moduleName: template-operator
  resources:
  - link: https://github.com/kyma-project/template-operator/releases/download/0.0.1/template-operator.yaml
    name: rawManifest
  version: 0.0.1
```

## Additional Resources

> - [`modulectl` GitHub Repository](https://github.com/kyma-project/modulectl)

Reviewer: ```suggestion
- [`/modulectl` GitHub Repository](https://github.com/kyma-project/modulectl)
```
